File: 2266_count_number_of_texts/Solution.cs
// LeetCode 2266 - Count Number of Texts
// https://leetcode.com/problems/count-number-of-texts/

public class Solution {
    public int CountTexts(string pressedKeys) {
        const int mod = 1000000007;
        int n = pressedKeys.Length;
        int[] dp = new int[n + 1];
        dp[0] = 1;
        for (int i = 1; i <= n; i++) {
            dp[i] = dp[i - 1];
            int maxPress = (pressedKeys[i - 1] == '7' || pressedKeys[i - 1] == '9') ? 4 : 3;
            for (int j = 2; j <= maxPress && j <= i; j++) {
                if (pressedKeys[i - j] != pressedKeys[i - 1]) break;
                dp[i] = (dp[i] + dp[i - j]) % mod;
            }
        }
        return dp[n];
    }
}

File: 2266_count_number_of_texts/solution.rs
// LeetCode 2266 - Count Number of Texts
// https://leetcode.com/problems/count-number-of-texts/

impl Solution {
    pub fn count_texts(pressed_keys: String) -> i32 {
        const MOD: i32 = 1_000_000_007;
        let s = pressed_keys.as_bytes();
        let n = s.len();
        let mut dp = vec![0i32; n + 1];
        dp[0] = 1;
        for i in 1..=n {
            dp[i] = dp[i - 1];
            let max_press = if s[i - 1] == b'7' || s[i - 1] == b'9' { 4 } else { 3 };
            for j in 2..=max_press {
                if j > i {
                    break;
                }
                if s[i - j] != s[i - 1] {
                    break;
                }
                dp[i] = (dp[i] + dp[i - j]) % MOD;
            }
        }
        dp[n]
    }
}
